### DISCUSSION TITLE
Pre-process POST request to CRUD to handle multiple data buckets

### DISCUSSION THREAD

**Discussion by LukMas** [mockoon/mockoon#1417] [Ideas] [answered]
Hi!

I've would know if it would be possible to add a pre-process step to the POSTs requests to CRUD (or handle directly them, like for the response) to handle multiple data buckets and not just one. 

I explain the things in detail. 
We are modeling our API according to JSON:API rules. While generating the responses requires some work, but it's doable, managing the POSTs is, in some cases, not possible. Or, while it's possible, it invalidates the whole mock data because a post of a resource could cause the update of multiple data buckets, but actually this is not possible. 
The simplest resource in JSON:API is something like this:

```
{
  "data": [{
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON:API paints my bikeshed!",
      "body": "The shortest article. Ever.",
      "created": "2015-05-22T14:56:29.000Z",
      "updated": "2015-05-22T14:56:28.000Z"
    },
    "relationships": {
      "author": {
        "data": {"id": "42", "type": "people"}
      }
    }
  }]
}
```
The `included` section is not listed to don't make the things even more complex. 
`Relationships` is present only if the resource has some relationships with others (in this case, with `people` in the form of `author` of the `article`).

A GET of such a resource requires some work in the templating, but it's doable (to a kind of degree - included not present). 
A POST of a resource without `relationships` is also possible to manage. The json data is stored within the data bucket, an ID is generated if not set (good that the ID is a string, that solves a lot of headaches) and the resource is available. 
The problem is when the resource carries extra informations in the `relationship` attribute. Because, as the specification says, the resource should carry also this information if it's relevant for the resource that is going to be stored. That means, a post like this 
```
{
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "relationships": {
      "photographer": {
        "data": { "type": "people", "id": "9" }
      }
    }
  }
}
```
is legitimate (as explained [here](https://jsonapi.org/format/#crud)). 
While with the actual management of the CRUD endpoint this POST updates correctly the resource in the data bucket, there's no way to update the data bucket that stores the relationship defined between `photo` and `people`. Hence, for relationship that are not just 1:1, there is no way to obtain the correct list of related resources because the data is not stored in the data bucket created for them.

I understand that this is a mocking service. That means, this feature goes a bit too far and maybe even beyond the real usage area of Mockoon. Anyway, to a kind of degree it would be nice to allows a customization, or manipulation, of the content of the request before interacting directly with the data bucket.

**Comment by 255kb** [comment, marked as answer]
Relations between data will probably be added in the future, but it may be simpler, a bit like what [json-server](https://github.com/typicode/json-server) is offering.
I understand the value of what you are proposing, but I'm not sure we will be able to support all the specifications like JSON:API.

**Reply by LukMas** [reply]
[original text removed because it was related to another answer]
Ok, I understand. It's a mocking service, hence it is not going to simulate a whole db (otherwise it loses its reason to exist). 

I suppose that a way to customize the Id or allow to modify the response would be already enough. I could also prepare a set of data, and if I can customize the response I could use some part of this set to simulate new elements. I've to investigate better this. 

Thank you!
